dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ConstantPropagation.java

```diff
@@ -1,313 +1,264 @@
 package com.dat3m.dartagnan.program.processing;
 
 import com.dat3m.dartagnan.expression.*;
-import com.dat3m.dartagnan.expression.op.BOpUn;
 import com.dat3m.dartagnan.expression.op.IOpUn;
-import com.dat3m.dartagnan.expression.processing.ExprSimplifier;
+import com.dat3m.dartagnan.expression.processing.ExprTransformer;
 import com.dat3m.dartagnan.expression.processing.ExpressionVisitor;
 import com.dat3m.dartagnan.program.Program;
 import com.dat3m.dartagnan.program.Register;
 import com.dat3m.dartagnan.program.Thread;
-import com.dat3m.dartagnan.program.event.Tag;
-import com.dat3m.dartagnan.program.event.arch.lisa.RMW;
-import com.dat3m.dartagnan.program.event.arch.tso.Xchg;
 import com.dat3m.dartagnan.program.event.core.*;
 import com.dat3m.dartagnan.program.event.core.utils.RegWriter;
-import com.dat3m.dartagnan.program.event.lang.catomic.AtomicCmpXchg;
-import com.dat3m.dartagnan.program.event.lang.catomic.AtomicLoad;
-import com.dat3m.dartagnan.program.event.lang.llvm.LlvmLoad;
 import com.dat3m.dartagnan.program.event.visitors.EventVisitor;
 import com.google.common.base.Preconditions;
-import com.google.common.base.Verify;
+import com.google.common.base.Predicate;
 import org.sosy_lab.common.configuration.Configuration;
 import org.sosy_lab.common.configuration.InvalidConfigurationException;
+import org.sosy_lab.common.configuration.Option;
+import org.sosy_lab.common.configuration.Options;
 
 import java.math.BigInteger;
-import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 
-import static com.dat3m.dartagnan.expression.op.IOpUn.BV2INT;
-import static com.dat3m.dartagnan.expression.op.IOpUn.BV2UINT;
+import static com.dat3m.dartagnan.configuration.OptionNames.CONSTANT_PROPAGATION;
+import static com.dat3m.dartagnan.configuration.OptionNames.PROPAGATE_COPY_ASSIGNMENTS;
 
+@Options
 public class ConstantPropagation implements ProgramProcessor {
-	
-    // =========================== Configurables ===========================
-
-    // =====================================================================
-
-	// Represents the top element of our lattice
-	private static final IConst TOP = new IConst() {
-		@Override
-		public BigInteger getValue() { return BigInteger.ZERO; }
-		@Override
-		public String toString() { return "T"; }
-		@Override
-		public int getPrecision() { return -1;}
-		@Override
-		public <T> T visit(ExpressionVisitor<T> visitor) { throw new UnsupportedOperationException(); }
-	};
 
-    private ConstantPropagation() { }
+    @Option(name = PROPAGATE_COPY_ASSIGNMENTS,
+            description = "Propagates copy assignments of the form 'reg2 := reg1' to eliminate " +
+                    "intermediate register 'reg2'. Can only be active if " + CONSTANT_PROPAGATION + " is enabled.",
+            secure = true)
+    private boolean propagateCopyAssignments = false;
```

Comment:
Why do we want to have fine grain control over this?

Reply:
It is not clear if this improves or degrades performance, so I didn't enable it by default.
It may be helpful to reduce the analysis time of the alias analysis though.

Reply:
The alias analysis time can change +-10% (in rt_mutex) even when it is run using exactly the same options. Enabling or not this option falls in the same range and thus I would consider it noise. For the time being, I would enable this option since it is the one that allows more simplifications in CP.